File: frontend/src/lib/engine/calibration.ts
/**
 * @file calibration.ts
 * @description Optical calibration engine implementing physical dimension measurement for Legal Metrology packaging.
 * Resolves the fundamental computer vision challenge of converting 2D digital image pixels to physical millimeters
 * using the internationally standardized GS1 EAN-13 barcode as a physical fiducial reference.
 */

import { PhysicalCalibrationMetrics } from '@/types/lmpc';

/**
 * Standard nominal physical dimensions of an EAN-13 barcode as specified by GS1 specifications.
 * At 100% nominal magnification, the barcode symbol (including quiet zones) measures:
 * Nominal Width: 37.29 mm
 * Nominal Height: 25.93 mm
 */
export const GS1_EAN13_NOMINAL_WIDTH_MM = 37.29;
export const GS1_EAN13_NOMINAL_HEIGHT_MM = 25.93;

/**
 * Rule 7 Table-I minimum numeral height thresholds in millimeters based on Principal Display Panel (PDP) area.
 * As prescribed in Legal Metrology (Packaged Commodities) Rules, 2011.
 *
 * | Area of PDP (cm²) | Minimum Height (mm) |
 * | :---              | :---                |
 * | A <= 50           | 1.0 mm              |
 * | 50 < A <= 100     | 1.5 mm              |
 * | 100 < A <= 500    | 2.5 mm              |
 * | A > 500           | 4.0 mm              |
 */
export const RULE_7_TABLE_I_THRESHOLDS = [
  { maxAreaCm2: 50, minHeightMm: 1.0 },
  { maxAreaCm2: 100, minHeightMm: 1.5 },
  { maxAreaCm2: 500, minHeightMm: 2.5 },
  { maxAreaCm2: Infinity, minHeightMm: 4.0 },
] as const;

/**
 * Calculates the statutory minimum numeral height required for a given Principal Display Panel area.
 *
 * @param pdpAreaCm2 The total surface area of the Principal Display Panel in square centimeters.
 * @returns The mandated minimum numeral height in millimeters under Rule 7 Table-I.
 */
export function getMandatedMinNumeralHeight(pdpAreaCm2: number): number {
  if (pdpAreaCm2 <= 0) return 1.0;
  for (const tier of RULE_7_TABLE_I_THRESHOLDS) {
    if (pdpAreaCm2 <= tier.maxAreaCm2) {
      return tier.minHeightMm;
    }
  }
  return 4.0;
}

/**
 * Performs optical calibration by converting pixel dimensions to physical millimeters using the detected barcode bounding box.
 *
 * @param barcodeWidthPx Width of the barcode detected in the image in pixels.
 * @param barcodeHeightPx Height of the barcode detected in the image in pixels.
 * @param numeralHeightPx Measured height of declaration characters/numerals in pixels.
 * @param numeralWidthPx Measured width of declaration characters/numerals in pixels.
 * @param imageWidthPx Total width of the captured package image in pixels.
 * @param imageHeightPx Total height of the captured package image in pixels.
 * @param customPdpAreaCm2 Optional estimated PDP area in cm²; if omitted, estimated from image proportions.
 * @returns Complete physical calibration and Rule 7 compliance evaluation metrics.
 */
export function calibratePhysicalMetrics(
  barcodeWidthPx: number,
  barcodeHeightPx: number,
  numeralHeightPx: number,
  numeralWidthPx: number,
  imageWidthPx: number = 1200,
  imageHeightPx: number = 1600,
  customPdpAreaCm2?: number
): PhysicalCalibrationMetrics {
  // Guard against division by zero or invalid negative numbers
  const safeBarcodeWidth = Math.max(10, barcodeWidthPx);
  const safeBarcodeHeight = Math.max(10, barcodeHeightPx);

  // Derive optical scale: Millimeters per pixel on the package plane
  const mmPerPixel = GS1_EAN13_NOMINAL_WIDTH_MM / safeBarcodeWidth;

  // Measure physical dimensions of the text numerals
  const measuredNumeralHeightMm = Number((numeralHeightPx * mmPerPixel).toFixed(2));
  const measuredNumeralWidthMm = Number((numeralWidthPx * mmPerPixel).toFixed(2));

  // Determine Principal Display Panel area
  let pdpAreaCm2 = 0;
  if (customPdpAreaCm2 && customPdpAreaCm2 > 0) {
    pdpAreaCm2 = customPdpAreaCm2;
  } else {
    // Standard estimation: compute total image physical area in cm²
    const imgWidthMm = imageWidthPx * mmPerPixel;
    const imgHeightMm = imageHeightPx * mmPerPixel;
    // Assuming package face occupies ~75% of camera view frame
    const estimatedFaceAreaCm2 = (imgWidthMm * imgHeightMm * 0.75) / 100;
    pdpAreaCm2 = Number(Math.max(20, Math.min(800, estimatedFaceAreaCm2)).toFixed(1));
  }

  // Lookup minimum mandated height per Rule 7 Table-I
  const mandatedMinHeightMm = getMandatedMinNumeralHeight(pdpAreaCm2);

  // Evaluate Rule 7 height constraint
  const isFontHeightCompliant = measuredNumeralHeightMm >= mandatedMinHeightMm;

  // Evaluate Rule 7 width constraint (width must not be less than 1/3 height, except for numeral '1' and letter 'I')
  const widthToHeightRatio = measuredNumeralHeightMm > 0
    ? Number((measuredNumeralWidthMm / measuredNumeralHeightMm).toFixed(2))
    : 0;

  return {
    barcodeWidthPx: safeBarcodeWidth,
    barcodeHeightPx: safeBarcodeHeight,
    mmPerPixel: Number(mmPerPixel.toFixed(4)),
    pdpAreaCm2,
    measuredNumeralHeightMm,
    measuredNumeralWidthMm,
    mandatedMinHeightMm,
    isFontHeightCompliant,
    widthToHeightRatio,
  };
}

File: backend/app/engine/calibration.py
"""
@file calibration.py
@description Optical fiducial scale calibration and Rule 7 Table-I font geometry verification.
Calibrates camera pixel coordinates to physical millimeters using the standard GS1 EAN-13 barcode fiducial (37.29 mm nominal width).
"""

from app.schemas import PhysicalCalibrationMetrics

# Standard GS1 General Specifications nominal dimensions for EAN-13 Barcode (100% magnification)
NOMINAL_EAN13_WIDTH_MM = 37.29
NOMINAL_EAN13_HEIGHT_MM = 25.93


def get_mandated_font_height_mm(pdp_area_cm2: float, is_blown_or_moulded: bool = False) -> float:
    """
    Computes mandatory minimum numeral height based on Principal Display Panel (PDP) area
    under Legal Metrology (Packaged Commodities) Rules, 2011 - Rule 7, Table-I.

    Table-I Statutory Thresholds:
    1. Area <= 50 cm²: Normal: 1.0 mm, Blown/Moulded: 1.5 mm
    2. 50 cm² < Area <= 100 cm²: Normal: 1.5 mm, Blown/Moulded: 3.0 mm
    3. 100 cm² < Area <= 500 cm²: Normal: 2.0 mm, Blown/Moulded: 4.0 mm
    4. 500 cm² < Area <= 2500 cm²: Normal: 4.0 mm, Blown/Moulded: 6.0 mm
    5. Area > 2500 cm²: Normal: 6.0 mm, Blown/Moulded: 6.0 mm
    """
    if pdp_area_cm2 <= 50.0:
        return 1.5 if is_blown_or_moulded else 1.0
    elif pdp_area_cm2 <= 100.0:
        return 3.0 if is_blown_or_moulded else 1.5
    elif pdp_area_cm2 <= 500.0:
        return 4.0 if is_blown_or_moulded else 2.0
    elif pdp_area_cm2 <= 2500.0:
        return 6.0 if is_blown_or_moulded else 4.0
    else:
        return 6.0


def calibrate_optical_metrics(
    barcode_width_px: float | None = None,
    barcode_height_px: float | None = None,
    numeral_box_height_px: float | None = None,
    numeral_box_width_px: float | None = None,
    package_type: str = "RECTANGULAR",
    package_width_cm: float = 7.5,
    package_height_cm: float = 16.0,
    is_blown_or_moulded: bool = False,
    image_width_px: int = 1280,
) -> PhysicalCalibrationMetrics:
    """
    Derives physical dimensions (in mm) from image pixel bounding boxes using optical fiducial scaling.
    If barcode width is provided, uses 37.29 mm / barcode_width_px.
    Otherwise estimates mm_per_pixel based on typical smartphone focal distance / sensor width.
    """
    if barcode_width_px and barcode_width_px > 10.0:
        mm_per_pixel = NOMINAL_EAN13_WIDTH_MM / barcode_width_px
        b_width_px = float(barcode_width_px)
        b_height_px = (
            float(barcode_height_px)
            if barcode_height_px
            else (b_width_px * (NOMINAL_EAN13_HEIGHT_MM / NOMINAL_EAN13_WIDTH_MM))
        )
    else:
        # Standard calibration fallback for 1080p-1440p smartphone cameras at ~25cm distance
        # Typical FOV at 25cm is ~200mm across the sensor width
        mm_per_pixel = 200.0 / max(float(image_width_px), 800.0)
        b_width_px = NOMINAL_EAN13_WIDTH_MM / mm_per_pixel
        b_height_px = NOMINAL_EAN13_HEIGHT_MM / mm_per_pixel

    # Calculate Principal Display Panel Area per Rule 2(h)
    if package_type.upper() in ["CYLINDRICAL", "TUBE", "BOTTLE"]:
        # For cylindrical/tube container: 40% of height x circumference (or 40% of height x width x pi)
        # Using projected frontal dimensions: 0.4 * height * (width * 3.14159)
        pdp_area_cm2 = 0.40 * package_height_cm * (package_width_cm * 3.14159)
    else:
        # Standard rectangular carton/pouch: Height x Width
        pdp_area_cm2 = package_height_cm * package_width_cm

    mandated_min_h = get_mandated_font_height_mm(pdp_area_cm2, is_blown_or_moulded)

    # Measured font height in mm
    if numeral_box_height_px and numeral_box_height_px > 0:
        measured_numeral_height_mm = round(numeral_box_height_px * mm_per_pixel, 2)
    else:
        # Default typical measured size if no specific box provided
        measured_numeral_height_mm = round(22.0 * mm_per_pixel, 2)

    if numeral_box_width_px and numeral_box_width_px > 0:
        measured_numeral_width_mm = round(numeral_box_width_px * mm_per_pixel, 2)
    else:
        measured_numeral_width_mm = round(measured_numeral_height_mm * 0.45, 2)

    width_ratio = round(measured_numeral_width_mm / max(measured_numeral_height_mm, 0.1), 2)
    is_compliant = (measured_numeral_height_mm >= mandated_min_h) and (width_ratio >= 0.33)

    return PhysicalCalibrationMetrics(
        barcode_width_px=round(b_width_px, 1),
        barcode_height_px=round(b_height_px, 1),
        mm_per_pixel=round(mm_per_pixel, 4),
        pdp_area_cm2=round(pdp_area_cm2, 2),
        measured_numeral_height_mm=measured_numeral_height_mm,
        measured_numeral_width_mm=measured_numeral_width_mm,
        mandated_min_height_mm=mandated_min_h,
        is_font_height_compliant=is_compliant,
        width_to_height_ratio=width_ratio,
    )
